### PR TITLE
configurable shutdownTimeout for Kakfa supervisor

### LINKED ISSUE #3495: Graceful shutdown of kafka supervisor does not work properly
In case of graceful shutdown `workerExec` executor service is used to get the highest offset of tasks in the task group, set that as end offset for all tasks and resume them so that they finish and publish. However, since in most cases doing this will take more time than the time out for shutting down supervisor i.e. `SHUTDOWN_TIMEOUT_MILLIS`(15 seconds), `shutdownNow()` will be called on `workerExec`. Thus, when all the futures for pausing tasks resolves, further logic for getting highest offset and resuming them will not run and fail with exception like -
```
Sep 21, 2016 9:42:33 PM com.google.common.util.concurrent.ExecutionList executeListener
SEVERE: RuntimeException while executing runnable com.google.common.util.concurrent.Futures$ChainingListenableFuture@5fde229f with executor com.google.common.
util.concurrent.MoreExecutors$ListeningDecorator@6c5b645
java.util.concurrent.RejectedExecutionException: Task com.google.common.util.concurrent.Futures$ChainingListenableFuture@5fde229f rejected from java.util.conc
urrent.ThreadPoolExecutor@2fde4963[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 10]
        at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2047)
        at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:823)
        at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1369)
        at com.google.common.util.concurrent.MoreExecutors$ListeningDecorator.execute(MoreExecutors.java:484)
        at com.google.common.util.concurrent.ExecutionList.executeListener(ExecutionList.java:156)
        at com.google.common.util.concurrent.ExecutionList.execute(ExecutionList.java:145)
        at com.google.common.util.concurrent.AbstractFuture.set(AbstractFuture.java:185)
        at com.google.common.util.concurrent.Futures$CombinedFuture.setOneValue(Futures.java:1626)
        at com.google.common.util.concurrent.Futures$CombinedFuture.access$400(Futures.java:1470)
        at com.google.common.util.concurrent.Futures$CombinedFuture$2.run(Futures.java:1548)
        at com.google.common.util.concurrent.MoreExecutors$SameThreadExecutorService.execute(MoreExecutors.java:297)
        at com.google.common.util.concurrent.ExecutionList.executeListener(ExecutionList.java:156)
        at com.google.common.util.concurrent.ExecutionList.execute(ExecutionList.java:145)
        at com.google.common.util.concurrent.ListenableFutureTask.done(ListenableFutureTask.java:91)
        at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:384)
        at java.util.concurrent.FutureTask.setException(FutureTask.java:251)
        at java.util.concurrent.FutureTask.run(FutureTask.java:271)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
```
The paused tasks will do nothing and will never be finish.

### PR DESCRIPTION
**Pull request by dclim**
Allow the shutdownTimeout to be configured if tasks are taking a long time to gracefully shutdown. Also increased the default from 15 seconds to 80 seconds which is greater than the default task client retry period which should give tasks lots of time to respond to the pause/publish commands.

Fixes #3495

### CODE REVIEW

**Comment by dclim** [comment]
@pjain1 let me know if this works for you guys

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java [hunk 1]
```diff
@@ -365,11 +364,12 @@ public void stop(boolean stopGracefully)
...
           }
 
-          long endTime = System.currentTimeMillis() + SHUTDOWN_TIMEOUT_MILLIS;
+          long shutdownTimeoutMillis = tuningConfig.getShutdownTimeout().getMillis();
+          long endTime = System.currentTimeMillis() + shutdownTimeoutMillis;
           while (!stopped) {
             long sleepTime = endTime - System.currentTimeMillis();
             if (sleepTime <= 0) {
-              log.info("Timed out while waiting for shutdown");
+              log.info("Timed out while waiting for shutdown (timeout [%dms])", shutdownTimeoutMillis);
```
Comment (gianm) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java]: [nit] `%,d` is more readable for potentially long values.

**Comment by dclim** [comment]
hey @pjain1, if you have a chance to review this we can try to get it into the 0.9.2 RC

**Comment by pjain1** [comment]
@dclim configurable timeout looks good to me however I was wondering if we need to kill the tasks in case of a timeout as it may happen that supervisor may not get change to resume the tasks and those tasks will never finish. What do you think ?

**Comment by dclim** [comment]
@pjain1 that makes sense to me, probably if they haven't responded by the timeout they're not in the best of states and we don't expose the chat handler endpoints for the task so all the user would be able to do anyway is to kill it manually; @gianm any concerns?

**Comment by gianm** [comment]
Killing tasks that the supervisor believes to be in a bad state is probably a good idea.

The worst that can happen is a rogue supervisor killing tasks inappropriately, although we could address that in the future with some epoch stuff. Even without that, though, I think the benefits of killing nonresponsive tasks outweigh the risks.

**Comment by gianm** [comment]
btw, by "some epoch stuff" I mean #3498.

**Comment by dclim** [comment]
@pjain1 to kill the tasks properly after the timeout (basically logic to avoid killing tasks that have already started publishing) requires a little more invasive changes than I first thought; could I post those changes in another PR so that this one can get in sooner?

**Comment by gianm** [comment]
@dclim IMO it's fine to do that in a different PR

**Comment by gianm** [comment]
btw, @dclim, your comment about avoiding killing tasks that are publishing makes me wonder: what kind of supervisor shutdown are we talking about here?

If it's a shutdown that is actually a restart, where we expect the supervisor to come back at some point, then it's probably better to not kill tasks at all. The new supervisor can do that if it wants to. If it's a shutdown that is meant to be final, I'm comfortable with killing tasks that are publishing. We want ingestion to stop, and it shouldn't really be useful to let publishing tasks finish.

So it seems to me that there may not be any cases where we want to treat publishing tasks specially. Unless I'm missing something.

**Comment by dclim** [comment]
@gianm hm, restarts / shutdowns currently use the same logic. 

If the shutdown is a 'final' shutdown, I think there's value in letting running tasks finish and publish. That would allow someone to use shutdown as a way to do an early stop without waiting for taskDuration to elapse - i.e. they spin up a supervisor, pump 1000 messages into it, and then shutdown the supervisor to publish those segments regardless of what taskDuration was set to.

Although maybe your point is more that if some tasks are in a bad state and they don't **all** publish their segments (therefore triggering the shutdown timeout), then having incomplete output is the same or worse than having no output at all and it doesn't matter if we kill the publishing tasks in this case - which I agree with.

**Comment by gianm** [comment]
I think it's a mistake, API-wise, to conflate "shutdown" and "flush". If we want to satisfy that use case you mentioned, we could introduce a "flush" API. This has a few advantages over combining them:

- callers that want to shutdown quickly without flushing data can just call "shutdown", and all inflight data will be dropped
- callers that want to flush data and then shutdown supervising can call "flush" followed by "shutdown" after all the data is flushed (which I guess the caller would check with some other API).
- avoids the awkward question of: if someone calls "shutdown" and that waits for tasks to publish, what happens if a task fails? Really, we should respawn it, or else this isn't a very useful flush API. But respawning tasks during shutdown mode is a bit strange and can delay shutdown by quite some time…

**Comment by gianm** [comment]
Similarly it's probably also a mistake to conflate "shutdown forevers" and "shutdown for the purposes of restarting soon". For the first probably the best experience is to nuke all inflight data, for the latter we definitely don't want to do that.

**Comment by dclim** [comment]
@gianm that API sounds good to me, I'll write it up as a separate issue

**Comment by pjain1** [comment]
@dclim separate PR for these changes should be fine

**Comment by pjain1** [comment]
👍

**Pull request merged**